components/raftstore-v2/src/operation/command/mod.rs

```diff
@@ -147,19 +206,160 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
-        }
-        // Note that the `commit_index` and `commit_term` here may be used to
-        // forward the commit index after being restarted. So it must be less
-        // than or equal to persisted index.
-        let commit_index = cmp::min(
-            self.raft_group().raft.raft_log.committed,
-            self.raft_group().raft.raft_log.persisted,
-        );
-        let commit_term = self.raft_group().raft.raft_log.term(commit_index).unwrap();
-        // TODO: schedule apply task
+            entry_and_proposals = committed_entries.into_iter().map(|e| (e, vec![])).collect();
+        }
+        // Unlike v1, v2 doesn't need to persist commit index and commit term. The
+        // point of persist commit index/term of raft apply state is to recover commit
+        // index when the writes to raft engine is lost but writes to kv engine is
+        // persisted. But in v2, writes to raft engine must be persisted before
+        // memtables in kv engine is flushed.
+        let apply = CommittedEntries {
+            entry_and_proposals,
+        };
+        self.apply_scheduler()
+            .send(ApplyTask::CommittedEntries(apply));
+    }
+
+    pub fn on_apply_res(&mut self, apply_res: ApplyRes) {
+        if !self.serving() {
+            return;
+        }
+        // It must just applied a snapshot.
+        if apply_res.applied_index < self.entry_storage().first_index() {
+            // TODO: handle admin side effects like split/merge.
+            return;
+        }
+        self.raft_group_mut()
+            .advance_apply_to(apply_res.applied_index);
+        let is_leader = self.is_leader();
+        let entry_storage = self.entry_storage_mut();
+        entry_storage
+            .apply_state_mut()
+            .set_applied_index(apply_res.applied_index);
+        entry_storage.set_applied_term(apply_res.applied_term);
+        if !is_leader {
+            entry_storage.compact_entry_cache(apply_res.applied_index + 1);
+            // TODO: handle read.
+        } else {
+            // TODO: handle read.
+        }
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    #[inline]
+    pub async fn apply_committed_entries(&mut self, ce: CommittedEntries) {
+        for (e, ch) in ce.entry_and_proposals {
+            if let Some(wb) = self.write_batch_mut() {
+                wb.set_save_point();
+            }
+            let resp = match self.apply_entry(&e).await {
+                Ok(resp) => resp,
+                Err(e) => {
+                    if let Some(wb) = self.write_batch_mut() {
+                        wb.rollback_to_save_point().unwrap();
+                    }
+                    cmd_resp::new_error(e)
+                }
+            };
+            self.callbacks_mut().push((ch, resp));
+            // Flush may be triggerred in the middle, so always update the index and term.
+            self.set_apply_progress(e.index, e.term);
```

Comment:
Why the progress is updated even when `apply_entry` failed?

Reply:
Because it's applied. We only care about consistency instead of success.